Add snap packaging recipe

forslund opened the pull request:
The snap package uses the python plugin to install the package using the setup.py script. The version is extracted from the setup.py as well.

The package ships with an included "hey mycroft" model.

forslund committed:
Add snapcraft packaging recipe

Also adds a SNAPCRAFT.md with some info of the specifics of the snap

forslund committed:
Add hey-mycroft model to snap

krisgesling commented on snap/snapcraft.yaml:
```diff
@@ -0,0 +1,138 @@
+build-packages: [python3-dev, python3-setuptools, libtool, libffi-dev, libssl-dev, autoconf, build-essential]
+parts:
+  alsa-mixin:
+    after: []
```
Are these empty lists needed? 
Should be able to delete them

krisgesling commented on snap/snapcraft.yaml:
```diff
@@ -0,0 +1,138 @@
+build-packages: [python3-dev, python3-setuptools, libtool, libffi-dev, libssl-dev, autoconf, build-essential]
+parts:
+  alsa-mixin:
+    after: []
```
Are these empty lists needed or there to explicitly show it doesn't happen after anything else?

forslund replied:
I believe those are just there because the template I used had them. Will remove

forslund replied:
Removed

krisgesling commented on snap/snapcraft.yaml:
```diff
@@ -0,0 +1,138 @@
+name: mycroft-precise
+adopt-info: mycroft-precise
+summary: Mycroft Precise
```
Just in terms of how it displays in the Snap store. This will come directly under "Mycroft-Precise". So perhaps use the description for the summary
"A wake word listener from Mycroft AI"

forslund replied:
Done

krisgesling commented on snap/snapcraft.yaml:
```diff
@@ -0,0 +1,138 @@
+name: mycroft-precise
+adopt-info: mycroft-precise
+summary: Mycroft Precise
+description: Mycroft Precise, a wake word spotter.
```
A lightweight, simple-to-use, RNN wake word listener, including all the tools to train your own custom wake word.

Mycroft Precise monitors an audio stream ( usually a microphone ) and when it recognizes a specific phrase it triggers an event. For example, by default users of the Mycroft Voice Assistant are using a Precise model trained to spot the phrase "Hey Mycroft". When Precise recognizes this phrase it puts the Mycroft Voice Assistant into command mode performing speech recognition on whatever is next said by the person using the device. 

Mycroft Precise is fully open source and can be trained to recognize any short-phrase or sound,  from a name to a cough. 

The default "Hey Mycroft" model is included so to test it out, try:
`mycroft-precise.listen /snap/mycroft-precise/current/hey-mycroft/hey-mycroft.pb`

#### USAGE

Running the listener:
- `mycroft-precise` - Alias for mycroft-precise.engine
- `mycroft-precise.engine` - Run a model on raw audio data from stdin
- `mycroft-precise.listen` - Run a model on microphone audio input
- `mycroft-precise.listen-pocketsphinx` - Run the PocketSphinx listener

Data collection:
- `mycroft-precise.collect` - Record audio samples for use with Precise
- `mycroft-precise.add-noise` - Create a duplicate dataset with added noise

Training:
- `mycroft-precise.train` - Train a new model on a dataset
- `mycroft-precise.train-generated` - Train a model on infinitely generated batches
- `mycroft-precise.train-incremental` - Train a model to inhibit activation by marking false activations and retraining
- `mycroft-precise.train-optimize` - Use black box optimization to tune model hyperparameters
- `mycroft-precise.train-sampled` - Train a model, sampling data points with the highest loss from a larger dataset

Evaluation and analysis:
- `mycroft-precise.test` - Test a model against a dataset
- `mycroft-precise.test-pocketsphinx` - Test PocketSphinx against a dataset
- `mycroft-precise.eval` - Evaluate a list of models on a dataset
- `mycroft-precise.calc-threshold` - Update the threshold values of a model for a dataset to make the sensitivity more accurate and linear
- `mycroft-precise.graph` - Show ROC curves for a series of models
- `mycroft-precise.simulate` - Simulate listening to long chunks of audio to find unbiased false positive metrics

Model conversion:
- `mycroft-precise.convert` - Convert wake-word model from Keras to TensorFlow

krisgesling replied:
Ake mentioned this might get pulled in from elsewhere, but we need to update the docs where ever it gets pulled from, hence this adds to that rewrite.

forslund replied:
Also it might not be a good idea to use the setup.py description since that can use mark-down in a way that the snap doesn't seem to handle in the description field.

forslund replied:
Did this but limited the markdown usage a bit to retain readability on the commandline.

krisgesling commented on SNAPCRAFT.md:
```diff
@@ -0,0 +1,31 @@
+# Mycroft-Precise Snap package
+
+## Package details
+
+The package is farily standard package using the Snap [Python Plugin](https://snapcraft.io/docs/python-plugin).
+
+The package version is extracted from the setup.py script to remain consistent with the pip information. In addition a comparison with the repo is made to see if this is commit matches the latest release commit, if it doesn't match _-dev_ is applied to the version info.
+
+Since Precise uses pyaudio, ALSA is required for things to work. This requires the [alsa-mixin](https://snapcraft-alsa.readthedocs.io/en/latest/snapcraft_usage.html) and results in some minor issues that needs to be handled. Library conflicts may occur if audio packages are included as part of the main application part, making them appear twice in conflicting versions. This is the reason that lib ...
+
+The snap package also includes a the current pretrained "hey mycroft" model accessed through `/snap/mycroft-precise/current/hey-mycroft/hey-mycroft.pb`
+
+## Plugs
+Plugs allows connecting the snap to the rest of the system, without any specified plugs the application will run without being able to access the system outside of the snap container, a read-only file system dedicated to the application.
```
Some minor suggestions throughout here: 
```
## Package details

The package is a fairly standard configuration using the Snap [Python Plugin](https://snapcraft.io/docs/python-plugin).

The package version is extracted from the `setup.py` script to remain consistent with the pip information. In addition a comparison with the repo is made to see if this commit matches the latest release commit. If it doesn't match _-dev_ is appended to the version info.

Since Precise uses pyaudio, ALSA is required for the audio to work. This Snap uses the [alsa-mixin](https://snapcraft-alsa.readthedocs.io/en/latest/snapcraft_usage.html) and results in some minor issues that need to be handled. Library conflicts may occur if audio packages are included as part of the main application part, making them appear twice in conflicting versions. This is the reason that libportaudio2 and pulseaudio are in the stage-packages for the alsa-mixin part.

This package also includes the current pretrained "hey mycroft" model accessed through `/snap/mycroft-precise/current/hey-mycroft/hey-mycroft.pb`.

## Plugs
Plugs allow the Snap to connect to specific aspects of the host system. Without any specified plugs the application will run entirely within a read-only file system dedicated to the application. It will not be able to access the system, including audio devices, that are outside of the Snap container.
```

forslund replied:
Reworded according to this.

krisgesling commented on SNAPCRAFT.md:
```diff
@@ -0,0 +1,31 @@
+- audio-record: Access to the system's audio input devices
+- audio-playback: Access to the system's audio output devices
+
+## Entry points
```
I wonder if we document these well in the snapcraft description so they're not needed here.

forslund replied:
Removed this section.

krisgesling commented on snap/snapcraft.yaml:
```diff
@@ -0,0 +1,138 @@
+    python-version: python3
+    source: https://github.com/MycroftAI/mycroft-precise.git
+    source-type: git
+    source-branch: dev
```
I wonder if we should fix up the branches before we publish this, and by default source this from the future master branch.

forslund replied:
It should be from the dev to work as both development/unstable and release versions

When a release is triggered dev will either match the latest tag -> a stable version string is generated
or
dev is different from the latest tag -> a version with "-dev" appendix will be added to the version.

forslund replied:
Dropped this section

forslund committed:
Update SNAPCRAFT.md wording

forslund committed:
Add proper global meta data

This is heavily based upon a draft by Kris Gesling, any errors are of course my
own.

forslund commented:
Thanks for all the good feedback and improvements. Merging now.
